Help Pagination

gazmull opened the pull request:
[![Build Status](https://travis-ci.org/gazmull/Misaki.svg?branch=help-pagination)](https://travis-ci.org/gazmull/Misaki)
* **Slighly modified help command for PaginationEmbed utility compatible**
* **Added PaginationEmbed under `util`**
  * Extends MessageEmbed

### Commit `fd27e36`
* **Make use of PaginationEmbed as a base class. (`pagination/base/PaginationEmbed`)**
* **Renamed `old` PaginationEmbed -> FieldsEmbed (`pagination/FieldsEmbed`)**
  * Also known as `Leaderboard-embed/Dynamic Mode`
* **Addition of support for MessageEmbed[] (`pagination/Embeds`)**
  * Also known as `Static Mode`
  * Added more methods for customising all embeds' properties (this overrides whatever you have set on a specific method(s) used from each `MessageEmbed` before the `Pagination` constructor. Notable method to customise is the `embed colour`; see example below).

Documentation: http://gazmull.github.io/discord-paginationembed

Examples:
```js
const FieldsEmbedMode = require(`${process.cwd()}/util/pagination/FieldsEmbed`);

new FieldsEmbedMode()
  .setArray([ { name: "John Doe" }, { name: "Jane Doe" } ])
  .setAuthorisedUser(message.author)
  .setChannel(message.channel)
  .setElementsPerPage(1)
  .setPage(2)
  .showPageIndicator(false)
  .build();
```
See screenshots on my next comment.
```js
const EmbedsMode = require(`${process.cwd()}/util/pagination/Embeds`);
const { MessageEmbed } = require("discord.js");

const embeds = [];

for (let i = 0; i < 5; ++i)
  embeds.push(new MessageEmbed().addField('Page', i + 1));

const myImage = message.author.displayAvatarURL();

new EmbedsMode()
  .setArray(embeds)
  .setAuthorisedUser(message.author)
  .setChannel(message.channel)
  .showPageIndicator(true)
  .setPage(3)
  .setImage(myImage) // Methods here and below are for customising all embeds
  .setThumbnail(myImage)
  .setTitle('Test Title')
  .setDescription('Test Description')
  .setFooter('Test Footer Text')
  .setURL(myImage)
  .setColor(0xFF00AE)
  .addBlankField()
  .addField('Test Field 1', 'Test Field 1', true)
  .addField('Test Field 2', 'Test Field 2', true)
  .build();
```
![image](https://user-images.githubusercontent.com/32944712/37118454-41116cbe-228f-11e8-9878-f39db26316a1.png)

gazmull committed:
PaginationEmbed Implementation

AndrewJWin reviewed (commented):
Working proof?

gazmull commented:
Right, here are some tests. Just a side note the first two tests were before the addition of "delete button (🗑).
![image](https://user-images.githubusercontent.com/32944712/36746277-e0d7c392-1c2c-11e8-96f2-545655383bd7.png)

Unnecessary, but to verify specific command help dialogs still work:
![image](https://user-images.githubusercontent.com/32944712/36746450-51031a90-1c2d-11e8-870f-c73274a18777.png)

kyranet commented:
The concept looks very well to me, it's documented and the code can be reused in other places, plus it's very clean. +1 in this PR. By any chance, I'll review this PR tomorrow after I get some free time.

Requesting @YorkAARGH to assign me to this PR.

DevYukine commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+        .showPageIndicator(true)
+        .setTimeout(60 * 1000);
+      
+      const isPage = !isNaN(parseInt(type));
```
IsNaN does accepts number strings aswell no need to parseint here

DevYukine commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+          .setArray(output)
+          .formatField("Categories", el => el);
+
+        if (isPage) embed.setPage(parseInt(type));
```
you should make a variable for the number version of type since you parseInt it 3 x later on what is kinda redundant

DevYukine commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+      
+        const output = [];
+        let num = 0;
+        const pg = parseInt(page) && parseInt(page) <= Math.ceil(n / perpage) ? parseInt(page) : 1;
```
what i said in line 78

DevYukine commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
-        .addField("Command description", cm.help.description)
-        .addField("Command usage", `\`${cm.help.usage}\``)
-        .addField("Command aliases", cm.conf.aliases.length == 0 ? "None" : cm.conf.aliases.join(", ") );
+      return await embed.build();
```
return await is kinda useless since if you return out of a async await it will auto await Promises anyways

kyranet replied:
Correction: `return await` is kinda useless as [AsyncFunction](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncFunction)s return a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). In order to get the result, you'd await that promise, and if you returned another promise, it'd await that one aswell.

DevYukine commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+        throw err;
+      });
+
+      if (c.constructor === Error) throw c;
```
couldn't you use instanceof here? is better practise imo

DevYukine commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+        throw err;
+      });
+
+      if (c.constructor === Error) throw c;
```
what i said in line 514 *probaly*

kyranet commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
 
-      embed.setDescription(description)
-        .addField("Categories", output);
+    if (this.client.commands.has(type) || this.client.commands.some(command => command.conf.aliases.includes(type))) {
```
Could be reduced by doing

```javascript
const command = this.client.commands.get(type) || this.client.aliases.get(type);
if (command) {
    // Do the operations
}
```

kyranet commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+        const tip = `\nTo skip to a page for more categories:\n\tReact with ↗; or\n\tDo \`${message.settings.prefix}help <page-num>\``;
+        const categories = sorted.filter(c => !(level < 10 && c.help.category == "Owner") || !(c.help.category === "NSFW" && !message.channel.nsfw)).map(c => {
+          const cat = c.help.category.toProperCase();
+          if (currentCategory !== cat && isDefault) {
```
You could avoid this and the latter processing by filtering it inside the `sorted.filter()`.

kyranet commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+        if (isPage) embed.setPage(parseInt(type));
+      } else {
+        let n = 0;
+        sorted.forEach(c => {
```
You could use this, which achieves the same goal but it's more performant. Check docs [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).

```javascript
const lowerCaseCategory = type.toLowerCase();
const n = sorted.reduce((acc, cur) => c.help.category.toLowerCase() === lowerCaseCategory ? cur + 1 : cur, 0);
```

kyranet commented on commands/system/help.js:
```diff
@@ -16,67 +17,98 @@ class Help extends Command {
+        let num = 0;
+        const pg = parseInt(page) && parseInt(page) <= Math.ceil(n / perpage) ? parseInt(page) : 1;
+        for (const c of sorted.values()) {
+          if (c.help.category.toLowerCase() === type.toLowerCase()) {
```
You already filtered by category twice, you need to filter it once with all the filters.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+  /**
+   * Options for PaginationEmbed.clientMessage.
+   * @typedef {Object} ClientMessageOptions
+   * @prop {Object} [message=null] - The message object sent by the client, if there is any.
```
Should use `@property` for clearlyness.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+   * Options for PaginationEmbed.clientMessage.
+   * @typedef {Object} ClientMessageOptions
+   * @prop {Object} [message=null] - The message object sent by the client, if there is any.
+   * @prop {String} [content="Preparing..."] - The custom message content while preparing the embed.
```
The types `String`, `Boolean`, `Number` refer to their constructors rather than their primitives. Please change them to `string`, `boolean` and `number`, respectively.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+   * @param {PaginationOptions} [options={}] Options for pagination utility.
+   */
+  constructor(options = {}) {
+    if (typeof options !== "object") throw new Error("Cannot invoke Pagination class without an actual options object.");
```
You'd like to check if null, `typeof null;` returns `object`.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+
+    /**
+     * The authorised user to navigate the pages.
+     * @type {Object}
```
`@type {User}`.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+
+    /**
+     * The channel where to send the embed.
+     * @type {Object}
```
`@type {TextChannel}`.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+   * // Object as constructor.
+   * const PaginationEmbed = require("<utils>/PaginationEmbed");
+   *
+   * // ...
```
Unnecesary comment.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+  /**
+   * Set the authorised person to navigate the pages.
+   * @param {Object} [user=null] - The user object.
+   * @default null
```
This is unnecesary. `@param {User} [user=null]` already states it defaults to `null`. Same to the occurrences below

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+   * Deploys emoji reacts for the MessageEmbed.
+   * @private
+   * @protected
+   * @returns {Void} - void
```
Unnecesary `@returns`. If none is written, void is implied.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+      return (
+        m.author.id === user.id && (
+          (!isNaN(supposedPage) && supposedPage !== this.page && supposedPage >= 1 && supposedPage <= this.pages) ||
+					m.content.toLowerCase() === "cancel"
```
Wrong indentation in this line.

kyranet commented on util/PaginationEmbed.js:
```diff
@@ -0,0 +1,564 @@
+  }
+}
+
+module.exports = PaginationEmbed;
```
Missing newline at the end of the file.

kyranet reviewed (changes requested):
The help command ended up being very similar to #68, and some of my requested changes are applied here.

gazmull commented:
Yes, it's similar to #68's help command source code. I thought it wasn't alright for me to rework the whole command for this pagination implementation PR. 😀 

Fixed majority, if not all, of requested changes. Thanks for the review.

gazmull committed:
Follow-up Adjustments

DevYukine reviewed (approved):
lgtm still waiting for @kyranet last words

kyranet commented on util/pagination/package.json:
```diff
@@ -0,0 +1,24 @@
+{
```
Mind explaining me what is this `package.json` here?

gazmull replied:
This was an "overlooked" file after deciding to ditch out a supposedly another pagination class (non-embed, which was the file pagination.js). Same with /pagination.js below, I will remove them on the next push.

kyranet commented on util/pagination/pagination.js:
```diff
@@ -0,0 +1,7 @@
+class PaginationEmbed {
```
This class is useless as it's not constructable and it has no propery, you don't need this.

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+
+      this._loadPage(content);
+    } catch (c) {
+      prompt.delete().catch(err => {
```
Missed await, in this case this promise throws, this will throw an unhanded error. Use the following code instead of re-throwing. Same with all the occurrences that follow this code pattern.

```javascript
await prompt.delete();
```

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+
+      content = parseInt(content);
+
+      this._loadPage(content);
```
You can remove the assignment before and do this:

```javascript
this._loadPage(parseInt(content));
```

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+      const field = this.fields[i];
+
+      if (typeof field.value === "function")
+        embed.addField(field.name, this.elementList.map(field.value).join("\n"), field.inline);
```
```javascript
embed.addField(field.name, typeof field.value === "function"
	? this.elementList.map(field.value).join("\n")
	: field.value, field.inline);
```

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+  showPageIndicator(boolean = true) {
+    if (typeof boolean !== "boolean") throw new Error("showPageIndicator() only accepts boolean type.");
+
+    this.pageIndicator = boolean === true;
```
The variable `boolean` is type boolean, enforced by a typeof check. The ` === true` is unnecesary.

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+
+    if (!(!isNaN(param) || isString)) throw new Error("setPage() only accepts number/string type.");
+
+    const navigator = {
```
```js
const navigator = param === 'back'
	? this.page === 1 ? this.page : this.page - 1
	: this.page === this.pages ? this.pages : this.page + 1;
```

kyranet commented on util/pagination/fields.js:
```diff
@@ -0,0 +1,574 @@
+ * @extends {MessageEmbed}
+ */
+class PaginationFieldsEmbed extends MessageEmbed {
+  /**
```
Put a space between the class declaration and the typedef.

kyranet reviewed (changes requested):
The filenames should change, `util/pagination/embeds.js` -> `util/pagination/PaginationEmbeds.js` and `util/pagination/fields.js` -> `util/pagination/PaginationFieldsEmbed.js` to match their exported class name.

Some requested changes repeat across all files. By any chance, you'd like `PaginationEmbeds` to extend `PaginationFieldsEmbed`, vice versa, or with another class. As a lot of code are duplicated between both, specially the constructors and multiple methods.

gazmull committed:
Addition of Embeds Mode + Adjustments

gazmull commented:
I've renamed the classes to `Embeds` and `FieldsEmbeds` respectively instead to reduce length of their name, but they still retain their identity. Also I added a base class to reduce SLOC significantly (pre-added modified `MessageEmbed` methods to `Embeds`).

kyranet reviewed (approved):
LGTM!
